Create a Scalar DL backup and restoration guide (#152)
scalar-labs/scalar-kubernetes: docs/ScalarDLBackupCreationRestoration.md

# Scalar DL Backup Creation and Restoration

> Scalar DL backup should be created after the Scalar DL server is paused otherwise there is a chance for inconsistency. 

Reviewer: ```suggestion
```

Author: Thank you for your suggestion.
Above sentence removed cd32752

> This guide shows you how to create and restore Scalar DL data backup without any data inconsistency.

Reviewer: ```suggestion
This guide shows you how to create and restore Scalar DL backups without any data inconsistency.
```

Author: Thank you for your suggestion.
Suggestion integrated cd32752

## Prerequisites

> * Read the Guide on How to [Back up and Restore Databases Integrated with Scalar DB](https://github.com/scalar-labs/scalardb/blob/master/docs/backup-restore.md)

Reviewer: The hypertext is wrong. (should include `Guide on How to`)

Author: Thank you for your suggestion.
Suggestion integrated cd32752

> * Scalar DL must be deployed in the Kubernetes cluster

Reviewer: ```suggestion
* Scalar DL must be deployed in a Kubernetes cluster
```

Author: Thank you for your suggestion.
Suggestion integrated cd32752

> * Cosmos DB account must be created with the backup policy `continuous` if you are using Cosmos DB.

Reviewer: ```suggestion
* Cosmos DB account must be created with the backup policy `continuous` if you use Cosmos DB.
```

Author: Thank you for your suggestion.
Suggestion integrated cd32752

> * You must synchronize time on the ledger and auditor servers if you are using the audit service.

Reviewer: ```suggestion
* You must synchronize the clocks (moderately) between Ledger and Auditor servers by using clock synchronization mechanisms such as NTP if you use Auditor. 
```

Author: Thank you for your suggestion.
Suggestion integrated cd32752

## Backup Creation

This section shows how to create a transactionally-consistent backup for Scalar DL.

### Requirements

> * You must wait at least 10 sec after pausing to create a backup.

Reviewer: why 10 seconds?
It should be decided based on clock drift.

Author: Sorry for the inconvenience.
Requirements update like this
```
* You must wait at least 10 sec (based on clock drift) after pausing to create a backup.
    * You must be able to identify a common restore point between the pausing based on the clock drift of the Ledger and Auditor.
```

kind: Pod
metadata:
  name: scalaradminutils
  namespace: default
spec:
  containers:
  - name: scalaradminutils
    image: ghcr.io/scalar-labs/scalar-admin:1.0.0
    command:
      - sleep
      - "3600"
    imagePullPolicy: IfNotPresent
  restartPolicy: Always
```

Use that manifest to create a Pod:

```console
kubectl apply -f scalaradminutils.yaml
```

Verify pod status:

```console
kubectl get pods scalaradminutils
```

### Pause

Use the following command to pause the Scalar DL service for few seconds

```console
kubectl exec -i -t scalaradminutils -- ./bin/scalar-admin -c pause -s <SRV_Service_URL>
```
Note:-
* Cosmos DB and DynamoDB will create backup automatically.

### Unpause

Use the following command to unpause the Scalar DL service after at least 10 seconds

```console
kubectl exec -i -t scalaradminutils -- ./bin/scalar-admin -c unpause -s <SRV_Service_URL>
```

## Restore

This section shows how to restore transactionally-consistent backup for Scalar DL.

### Requirements

* You must use a middle value of the paused time as restore point.
* You must restore Scalar Ledger and Auditor tables with the same restore point if you are using the ledger and audit services.

> ### Cosmos DB

Reviewer: For each database specific actions, I don't think they're k8s specific so should they be described in the backup doc?

Author: Sorry for the inconvenience.
This section has been removed from this guide cd32752

        
        B. Create the backup of previously restored tables (A) using the [AWS official guide](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Backup.Tutorial.html#backup_console).
        
        C. Delete all scalar/auditor tables except the previously restored tables (A).
        
        D. Restore the previously created backup (B) using the actual scalar/auditor table name (previously deleted tables (C)) on the basis of [AWS official guide](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Restore.Tutorial.html#restoretable_console).
    
    * On the AWS Command Line Interface, You can restore all tables using following shell script.
        * Configure AWS CLI with the region of DynamoDB.
        ```console
        aws configure
        ```
        * Download the [Point In Time Restore](./script/restore_dynamo.sh) script.
        * Execute the script with restore point as UNIX time (UTC) and all table names
            * `time` must be without millisecond
            * All tables except `scalardb.metadata` must be passed to the restore script
        ```console
        #Example
        #For ledger restoration

>         ./restore_dynamo.sh --time 1613960773 coordinator.state scalar.asset scalar.asset_metadata scalar.certificate scalar.contract scalar.contract_class scalar.function

Reviewer: As a guide, let's focus on describing the steps.
The tool should be more properly created and tested.

Author: Sorry for the inconvenience.
This section has been removed from this guide

> # Scalar DL Backup Creation on Kubernetes

Reviewer: Backup only? The next sentence says `create and restore Scalar DL backups`, so they are already inconsistent.

Reviewer: Also, this doc doesn't even describe how to take a backup, rather it described how to pause and unpause Scalar Ledger pods in k8s. It should describe what it does correctly. Very misleading.
Moreover, it doesn't really describe Scalar DL specific things.

* What should be done to do backup and restore properly in Scalar DL? Any difference when using Sclar DL with only Ledger or with Ledger and Auditor?  In Ledger+Auditor case, pausing Ledger only is fine? If so, why?

Author: > Backup only? The next sentence says create and restore Scalar DL backups, so they are already inconsistent.

Yes, Now I feel there is inconsistency (document name also different).

But during the period of document creation and self-review, I thought that this guide mainly handles the backup creation so thought heading `Scalar DL Backup Creation on Kubernetes` is enough.

Author: >Also, this doc doesn't even describe how to take a backup, rather it described how to pause and unpause Scalar Ledger pods in k8s. It should describe what it does correctly. Very misleading.
Moreover, it doesn't really describe Scalar DL specific things.
What should be done to do backup and restore properly in Scalar DL? Any difference when using Sclar DL with only Ledger or with Ledger and Auditor? In Ledger+Auditor case, pausing Ledger only is fine? If so, why?

I tried to avoid redundancy. A detailed explanation of backup creation and restoration exists in [A Guide on How to Backup and Restore Data in Scalar DL](https://github.com/scalar-labs/scalardl/blob/master/docs/backup-restore.md) so I specified this document in Prerequisites.

I tried to explain how to create a transactionally consistent backup for Scalar DL on Kubernetes. I thought the continuation of  [A Guide on How to Backup and Restore Data in Scalar DL](https://github.com/scalar-labs/scalardl/blob/master/docs/backup-restore.md)  makes sense here.

>any difference when using Sclar DL with only Ledger or with Ledger and Auditor?

I have added a Requirement `You must identify a unique point-in-time recovery/restore point for the Ledger and Auditor if you use both the Ledger and Auditor.`, and I thought other details will be received from  [A Guide on How to Backup and Restore Data in Scalar DL](https://github.com/scalar-labs/scalardl/blob/master/docs/backup-restore.md) 

>In Ledger+Auditor case, pausing Ledger only is fine? 

In my understanding, yes

> This guide shows you how to create and restore Scalar DL backups on Cosmos DB and DynamoDB without data inconsistencies.

Reviewer: Does this doc specific for Cosmos and Dynamo?
The title doesn't say so and the contents don't look like so.

> ## Prerequisites

Reviewer: I think this doc is overall redundant with the backup and restore guide.
If it is saying specifically for k8s, talk about k8s-specific things mainly here.

Author: > I think this doc is overall redundant with the backup and restore guide.

Yes, I admit that.
There are a lot of concept repetitions.
I will update the document without any concept repetitions.

Author: >If it is saying specifically for Scalar DL and k8s, talk about Scalar DL and k8s specific things here.

Ok, I will update the document with above details

> * Read the Scalar DL backup creation and restoration guide, [see](https://github.com/scalar-labs/scalardl/blob/master/docs/backup-restore.md)

Reviewer: Weird to add a link to `see`.

Author: I added `see` because the file name is `A Guide on How to Backup and Restore Data in Scalar DL` so I thought `see` is good.

I will update it with proper wordings

* Scalar DL Ledger (and Auditor) must be deployed in the Kubernetes cluster
* Cosmos DB account must be created with the backup policy `continuous` if you use Cosmos DB
* DynamoDB tables must be created with point-in-time recovery (by default, scalardl-schema-loader enables PITR)

## Backup Creation

This section shows how to create a transactionally-consistent backup for Scalar DL.

### Requirements

> * You must enable point-in-time recovery/restore in the backend database.

Reviewer: Is it saying the same thing as above in a different way? 
Very redundant and should be united.

Author: Sorry, It is missed.
I missed it because the same idea described different sentences.

I will update the document without repeating the concept.

* You must identify a unique `point-in-time recovery/restore` point for the Ledger and Auditor if you use both the Ledger and Auditor. 

> ### Recommendations

Reviewer: Redundant. (it is described in the backup and restore guide)

Author: Sorry for the inconvenience.
I will update the guide without concept repetitions.

* You should pause Scalar DL services for enough time (e.g., 10 seconds)  to create point-in-time recovery/restore
* You should use mid-time of the paused duration as a restore point since clock synchronization cannot perfectly synchronize clocks between nodes

> ### Find SRV service URL

Reviewer: I don't think there is a term called `SRV service URL`.
It's an `SRV record` or `service record`.
Use defined terms.

Author: Sorry for the inconvenience.
It is an `SRV record`

I will update the document.

> This section will help you to find the SRV service URL with the help of [SRV records](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#srv-records).

Reviewer: We don't need `This section will help you to ...` thing for this small section.
This kind of thing is used only in the beginning of a big section.

Author: Thank you for your explanation.
I will update the document.
